pkg/network/common/egressip.go

```diff
@@ -621,30 +822,103 @@ func (eit *EgressIPTracker) allocateExistingEgressIPs(allocation map[string][]st
 	return removedEgressIPs
 }
 
+type sortedNamespaces []*namespaceEgress
+
+func (d sortedNamespaces) Len() int {
+	return len(d)
+}
+
+func (d sortedNamespaces) Swap(i, j int) {
+	d[i], d[j] = d[j], d[i]
+}
+
+// Sort namespaces by increasing order following the amount requested per
+// namespace, i.e: the namespace requesting the lowest amount of IPs will be
+// first.
+func (d sortedNamespaces) Less(i, j int) bool {
+	return len(d[i].requestedIPs) < len(d[j].requestedIPs)
+}
+
+// getSortedEgressIPs will return all egress IPs in increasing order. The IPs
+// belonging to the namespace with the lowest amount of requested IPs will be
+// first.
+func (eit *EgressIPTracker) getSortedEgressIPs() []*egressIPInfo {
+	sortedNamespaces := make(sortedNamespaces, 0, len(eit.namespacesByVNID))
+	for _, namespace := range eit.namespacesByVNID {
+		sortedNamespaces = append(sortedNamespaces, namespace)
+	}
+	sort.Sort(sortedNamespaces)
+	sortedEgressIPs := make([]*egressIPInfo, 0, len(eit.egressIPs))
+	for _, namespace := range sortedNamespaces {
+		for _, egressIP := range namespace.requestedIPs {
+			sortedEgressIPs = append(sortedEgressIPs, eit.egressIPs[egressIP])
+		}
+	}
+	return sortedEgressIPs
+}
+
+type sortedNodes []*nodeEgress
+
+func (d sortedNodes) Len() int {
+	return len(d)
+}
+
+func (d sortedNodes) Swap(i, j int) {
+	d[i], d[j] = d[j], d[i]
+}
+
+// Sort egress nodes by increasing order following the nodes capacity, i.e: the
+// nodes with the lowest capacity will be first.
+func (d sortedNodes) Less(i, j int) bool {
+	return d[i].capacity < d[j].capacity
+}
+
+func (eit *EgressIPTracker) getSortedNodes() []*nodeEgress {
+	sNodes := make(sortedNodes, 0, len(eit.nodes))
+	for _, node := range eit.nodes {
+		sNodes = append(sNodes, node)
+	}
+	sort.Sort(sNodes)
+	return sNodes
+}
+
 func (eit *EgressIPTracker) allocateNewEgressIPs(allocation map[string][]string, alreadyAllocated map[string]bool) {
-	// Allocate pending egress IPs that can only go to a single node
-	for egressIP, eip := range eit.egressIPs {
-		if alreadyAllocated[egressIP] {
+	// Allocate pending egress IPs that can only go to a single node. Given a
+	// capacity constraint allocation problem: we need to begin by trying to
+	// assign all IPs in the namespace with the lowest amount of requested
+	// egress IPs. Given that these egress IP will be assigned to the nodes with
```

Comment:
Yes, you're right. I overlooked this and the test case I wrote didn't cover this scenario. I will update the sorting algorithm and add a test case which covers this.

Reply:
Actually, I realized we need a combination of both of our strategies. Given the following scenario:

node1, capacity = 2
node2, capacity = 1

namespace1, egressIPs = [1,2]
namespace2, egressIPs = [3]

we will need a final sorted egress IP list of: [3,1,2], since 3 will be assigned to node1 first, 1 to node2 and finally 2 to node1

if our final sorted egress IP list is: [1,3,2] we won't be able to assign everything, since 1 will be assigned to node1 first, 3 to node2 and finally 2 can't be assigned anywhere because 1 was already assigned to node1.